espnet/nets/pytorch_backend/e2e_asr_transformer.py

```diff
@@ -224,7 +249,14 @@ def recognize(self, feat, recog_args, char_list=None, rnnlm=None, use_jit=False)
         TODO(karita): do not recompute previous attention for faster decoding
         '''
         self.eval()
-        feat = torch.as_tensor(feat).unsqueeze(0)
+        feat = to_torch_tensor(feat).unsqueeze(0)
+        # 0. Frontend
+        if self.frontend is not None:
```

Comment:
I do not like this. Remember our `recognize` method is already too complex.

How about this?
```python

# espnet.nets.pytorch_backend.frontend_asr.py
class FrontendASR(torch.nn.Module):
    def __init__(self, frontend, asr_model):
        super().__init__()
        assert isinstance(asr_model, ASRInterface)
        self.frontend = frontend
        self.asr_model = asr_model

    def forward(self, xs_pad, ilens, ys_pad):
        hs_pad, hlens, _ = self.frontend(to_torch_tensor(xs_pad), ilens)
        hs_pad, hlens = self.feature_transform(hs_pad, hlens)
        return self.asr_model(hs_pad, hlens, ys_pad)

    def recognize(self, feat, recog_args, char_list=None, rnnlm=None, **kwargs):
        self.eval()
        feat = to_torch_tensor(feat).unsqueeze(0)
        feat, _, _ = self.frontend(feat, [feat.size(1)])
        feat, _ = self.feature_transform(feat, [feat.size(1)])
        feat = feat.squeeze(0)
        return self.asr_model(feat, recog_args, char_list, rnnlm, **kwargs)

    def enfance(...):
        ...
```

Reply:
If `self.frontend.forward` applies `to_torch_tensor` inside itself, we can generalize this class between pytorch and chainer.

Reply:
I'm not sure what you propose. 
Does FrontendAsr correspond to the current `E2E`? I can't get  how to instantiate the class. Do you mean this?

```python
model = FrontendASR(FrondEnd(...), OriginalE2E(...))
```

Reply:
maybe you can add `--frontend-module espnet.nets.xxx:YYY` and asr.py can be changed like
``` python
model = dynamic_import(args.model_module)
...
if args.frontend_module is not None:
    frontend = dynamic_import(args.frontend_module)
    model = FrontendASR(frontend, model)
```

Reply:
O.K. @ShigekiKarita 's plan looks nice. 
I was also concerned about modularity of FrontEnd. 

I'll notify if I finished.